designs/closure-design.md

```diff
@@ -0,0 +1,192 @@
+# type error
+f: Callable[[int], int] = None  
+g: Callable[[bool], int] = None
+f = g
+```
+
+```python
+# `is` should work on the callable object
+# expect True
+f: Callable[[int], int] = None  
+f is None 
+```
+
+```python
+# parse and type check closure
+# type check 
+def getAdder(a:int) -> Callable[[int], int]:
+    def adder(b: int) -> int:
+        return a + b
+    return adder
+# the remaining tests all require this definition at top
+```
+
+```python
+# the returned function can be called
+# expect: 3
+getAdder(1)(2) 
+```
+
+```python
+# assign a closure to a global variable
+# pass type check
+f: Callable[[int], int] = None
+f = getAdder(1) 
+```
+
+```python
+# assign and call
+# expect 3
+f: Callable[[int], int] = None
+f = getAdder(1) 
+f(2) 
+```
+
+```python
+# the returned closures are different objects
+# expect false
+f: Callable[[int], int] = None
+f = getAdder(1) 
+g: Callable[[int], int] = None
+g = getAdder(1) 
+f is g
+```
+
+```python
+# pass the same closure around and use `is` operator
+# expect true
+f: Callable[[int], int] = None
+f = getAdder(1) 
+g: Callable[[int], int] = None
+g = f
+f is g
+```
+
+```python
+# lexical scoping
+# expect 3
+a: int = 10
+f: Callable[[int], int] = None
+f = getAdder(1) 
+f(2) 
+```
+
+## changes
+
+### AST & IR
+
+See the file diff in the PR for the changes to AST. We don't plan to change the
+IR. The diff to `ast.ts` is also provided below.
+
+We first added a `func` variant in the `Type` and the subtyping rules for
+functions (the implementation doesn't support subclass checks). To simplify the
+implementation, we will initialize functions with a `None` like in ChocoPy.
+
+For statements and expressions, we added AST support for global/nonlocal
+statements, nested function definitions, and lambda expressions.
+
+```diff
+diff --git a/ast.ts b/ast.ts
+index 1c40771..ffbc6a4 100644
+--- a/ast.ts
++++ b/ast.ts
+@@ -6,6 +6,7 @@ export type Type =
+   | {tag: "bool"}
+   | {tag: "none"}
+   | {tag: "class", name: string}
++  | {tag: "func"; args: Type[]; ret: Type }
+   | {tag: "either", left: Type, right: Type }
+ 
+ export type Parameter<A> = { name: string, type: Type }
+@@ -26,6 +27,9 @@ export type Stmt<A> =
+   | {  a?: A, tag: "field-assign", obj: Expr<A>, field: string, value: Expr<A> }
+   | {  a?: A, tag: "if", cond: Expr<A>, thn: Array<Stmt<A>>, els: Array<Stmt<A>> }
+   | {  a?: A, tag: "while", cond: Expr<A>, body: Array<Stmt<A>> }
++  | {  a?: A, tag: "closure", func: FunDef<A> }
++  | {  a?: A, tag: "nonlocal", vars: string[] }
++  | {  a?: A, tag: "global", vars: string[] }
+ 
+ export type Expr<A> =
+     {  a?: A, tag: "literal", value: Literal }
+@@ -38,6 +42,7 @@ export type Expr<A> =
+   | {  a?: A, tag: "lookup", obj: Expr<A>, field: string }
+   | {  a?: A, tag: "method-call", obj: Expr<A>, method: string, arguments: Array<Expr<A>> }
+   | {  a?: A, tag: "construct", name: string }
++  | {  a?: A, tag: "lambda", args: string[], body: Expr<A> }
```

Comment:
Implementing lambda seems a lot to do for week 1. A realistic goal would be to maybe think about implementation of non-local variables and then maybe making a first-class function with no non-local variables working (please also refer to the piazza post)

Reply:
We are not planning to implement lambda this week. It's part of the final design. We just want to change the sharing interface once so we don't need to change it every week.